### PR TITLE
Tiffany/edit customer

### PR DESCRIPTION
**Pull request by cnnmon**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

You can use the [edit customer page](http://localhost:3000/customers/customer/edit) to edit customers in Airtable and Redux & create customer records!

## How to review

[//]: # 'The order in which to review files and what to expect when testing locally'

Edit customer -> check airtable
http://localhost:3000/customers/customer/edit

## Next steps

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

When you press the back button after editing, you go back to the old version of someone's profile (the changes are still made though, so idk whether that's important). Also weird errors in node console after you edit but it still works?

### Screenshots

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @julianrkung

[//]: # 'This tags Julian as a default. Feel free to change, or add on anyone who you should be in on the conversation.'

### CODE REVIEW

File: src/screens/Customers/EditCustomer.tsx
Comment (julianrkung) [inline src/screens/Customers/EditCustomer.tsx]: This should probably just be `customerId` since there's at most 1 customerId being changed for a customer update. We can change that in a separate PR but let's put a TODO?

File: src/screens/Customers/EditCustomer.tsx
Comment (julianrkung) [inline src/screens/Customers/EditCustomer.tsx]: We probably don't actually need this since it's never displayed in the app (mostly just something for the admin to see via Airtable). We might be able to just create the customer update on the backend (and not create one on the frontend)

We could probably delete this from the interface also for now if that's an issue

File: src/lib/airtable/interface.ts
Comment (julianrkung) [inline src/lib/airtable/interface.ts]: I think we should make this a string since there's only 1 customer (might need to change some stuff in Airtable if it's foreign-key-many but that can be separate pr)

File: src/lib/redux/siteDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/siteDataSlice.ts]: Should this be a `customerRecord`?

File: src/screens/Customers/CustomerProfile.tsx
Comment (julianrkung) [inline src/screens/Customers/CustomerProfile.tsx]: We should be consistent in either passing the `customer` via `Link` or using `currentCustomer` in redux. We don't have to do it in this PR, if you want we can put a TODO (and tag me so I do it)

File: src/lib/redux/siteDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/siteDataSlice.ts]: If the customer we are editing is the same as the current customer (which might always be the case?), we need to update the `currentCustomer`'s object so that it also reflects the changes.

Comment (julianrkung) [reply src/lib/redux/siteDataSlice.ts]: Maybe the ^ should happen in `editCustomerInRedux`? Do not know, whatever you think is cleaner

File: src/lib/redux/siteDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/siteDataSlice.ts]: I think this might miss if the customer was created offline --> edited offline. Let's talk about how that should work offline, because we can handle it the same way we handle payments and meter readings

**Review by julianrkung — COMMENTED** [review]
Looks really good as a first iteration, put some comments down!

File: src/screens/Customers/EditCustomer.tsx
Comment (julianrkung) [inline src/screens/Customers/EditCustomer.tsx]: Actually, this `editCustomerInRedux` call might even be better placed inside of `editCustomer`. Then, we can abstract away the `editCustomer` function as BOTH a frontend change and a backend change and ensure they're consistent since there should never be a case where I update the frontend and not the backend and vice verase. 

This applies to `addCustomerToRedux` as well. Thoughts on this? @wangannie

File: src/screens/Customers/EditCustomer.tsx
Comment (cnnmon) [reply src/screens/Customers/EditCustomer.tsx]: I don't need to delete customerUpdates from interface for functionality because it's taking from the currentCustomer's data, but let me know if you want me to delete it to clean things up in the future!

File: src/lib/redux/siteDataSlice.ts
Comment (cnnmon) [reply src/lib/redux/siteDataSlice.ts]: Should currentSite be SiteRecord then?

File: src/lib/airtable/interface.ts
Comment (cnnmon) [reply src/lib/airtable/interface.ts]: I changed schemas to foreign-key-one to make sure everything worked here! What else do I need to do past that?

File: src/components/TextField.tsx
Comment (wangannie) [inline src/components/TextField.tsx:39]: Nit: empty string instead of undefined

File: src/lib/redux/siteDataSlice.ts
Comment (wangannie) [reply src/lib/redux/siteDataSlice.ts]: yes^

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx:37]: Not exactly related to this line/this probably isn't really relevant to your PR but it seems like `UserFields` is pretty out of date?

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx:47]: I see there's a `hasMeter` but not a way to edit the meter number -- if the customer used to have a meter number but was edited so that this is un-selected, it doesn't clear the meter number.

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx]: You should negate the `currentCustomer.isactive` -- it's displaying the opposite every time

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx]: Since it's Explanation in Airtable, wouldn't it make sense to keep it consistent throughout? Or just change it to Reason in Airtable?

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [reply src/screens/Customers/EditCustomer.tsx]: Ooh yes agreed.

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx]: Change to const

File: src/screens/Customers/CustomerProfile.tsx
Comment (wangannie) [inline src/screens/Customers/CustomerProfile.tsx]: You can resolve the weird node error if you wrap this in a hook:

```
  useEffect(() => {
    setCurrentCustomer(customer);
  },[]); 
```

Source: https://stackoverflow.com/questions/62336340/cannot-update-a-component-while-rendering-a-different-component-warning

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [inline src/screens/Customers/EditCustomer.tsx:120]: I'm confused why `currentCustomer` is part of `siteData` at all. Why is this necessary/why can't it just be passed around where needed? Also because it's not being reset/cleared when you  navigate away from the customer profile.

Comment (wangannie) [reply src/screens/Customers/EditCustomer.tsx:120]: Ok yeah I just saw Julian's note on this (https://github.com/calblueprint/meepanyar/pull/55/files#r598159199)

File: src/lib/redux/siteDataSlice.ts
Comment (wangannie) [inline src/lib/redux/siteDataSlice.ts]: This is a duplicate! It already exists in `customerUtils`

File: src/screens/Customers/EditCustomer.tsx
Comment (wangannie) [reply src/screens/Customers/EditCustomer.tsx:120]: Actually relying on `currentCustomer` instead of props for this screen could resolve your issue of having to go back twice (I spent a while digging into react router trying to come up with a fix for that but I couldn't find much).

**Review by wangannie — COMMENTED** [review]
Looking good!  Noted some cleanup stuff + some questions about the workflow. Overall, I'm really confused why currentCustomer is necessary/ whether it's a good idea to keep.

Also, in general, it would help if you added some more detail to your PR description! Specifically more detail on how to review (at least linking the backend PR) + giving more explicit steps of sample workflows (including edge cases) to test and results to expect. Detailed PR descriptions also serve as really good documentation to reference later on, which makes writing handoff documentation a lot easier!

Higher-level question: what's the point of making a customer update record? Besides adding the explanation, it isn't tracking anything, so I'm not sure what the utility is.

File: src/screens/Customers/EditCustomer.tsx
Comment (julianrkung) [reply src/screens/Customers/EditCustomer.tsx]: +1 should be consistent, thanks for catching

**Comment by julianrkung** [comment]
> Higher-level question: what's the point of making a customer update record? Besides adding the explanation, it isn't tracking anything, so I'm not sure what the utility is.

Good question. It's solely for the NPO to be able to see via the Airtable (I don't believe there are plans to access it in the app, but it's a paper trail)

File: src/lib/redux/siteDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/siteDataSlice.ts]: We can write the code as if the redux store is mutable. This is my bad, sorry.

File: src/screens/Customers/EditCustomer.tsx
Comment (cnnmon) [reply src/screens/Customers/EditCustomer.tsx:47]: So to make sure, people will generally be moved to new meter numbers?

File: src/components/TextField.tsx
Comment (cnnmon) [reply src/components/TextField.tsx:39]: if i put it as an empty string it doesn't let anyone type anything

File: src/screens/Customers/CustomerProfile.tsx
Comment (julianrkung) [inline src/screens/Customers/CustomerProfile.tsx:58]: Note to other reviewers: I'll make sure to get rid of this before merging https://github.com/calblueprint/meepanyar/pull/56 . I plan on merging the `customerData` refactor PR AFTER this editCustomer PR is merged

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: Note to other reviewers: This workflow is consistent with the offline workflow we have for payments and meter reading functionality. This'll probably need to be added to / adjusted in the future once we have more functionality set up, but I'm okay with approving this for now and dealing with it down the road.

File: src/screens/Customers/EditCustomer.tsx
Comment (julianrkung) [inline src/screens/Customers/EditCustomer.tsx]: is this TODO already done?

File: src/lib/redux/siteDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/siteDataSlice.ts:79]: Can you put a note that this doesn't currently work for customers created while offline since they don't have an ID? We can address it in a future PR but we should definitely put that down so we remember.

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: I'd be careful with the variable naming here. This object is a `customer` with an added `customerUpdate` in it. Therefore, when you do `editCustomerInRedux(customer)`, the edited customer you added now looks different than all the other customers (the edited customer has a `customerUpdate` field that no other customer has).

You could:
1. pass them as 2 separate parameters initially and when you do the fetch you can `JSON.stringify({...customer, customerUpdate})` like how you did when calling `editCustomer`)
2. If you look at the custom backend endpoint, [here](https://github.com/calblueprint/meepanyar-node/pull/15/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R125), we really just separate out the data and do a `updateCustomer` and `createCustomerUpdate`. I think we could instead just separate out the data here and call `updateCustomer` and `createCustomerUpdate` here and not need the custom backend endpoint. I'm not sure if I'm missing something with this solution though, so let me know if you have concerns. 

If solution 2 doesn't have problems, I'd go with that since the frontend has built-in error handling (thanks to the offline library we're using), so we'll retry in the case either our network goes down or the backend is down etc. We could also just go with solution #1 for now and revisit solution #2 later. Your call

**Review by julianrkung — COMMENTED** [review]
It's looking really good, I really only had the comment in `editCustomer`. I think after this iteration, we should be good to merge!

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: We also have to editCustomerInRedux after the if statement to update the redux store (maybe this just happens after the `if / else`?

**Review by julianrkung — APPROVED** [review]
Approved knowing of bug where you return to a previously un-edited profile. Will fix in a future PR. There's a branch where this is fixed: https://github.com/calblueprint/meepanyar/compare/tiffany%2Fedit-customer...annie%2Fedit-customer-suggestions?diff=split&expand=1 

that can fix the bug

**Pull request merged**
